### ISSUE TITLE
Decode accumulating on nested objects

### ISSUE THREAD

**Issue by dr4ke616** [circe/circe#1078]
Hey,

Is there a mechanims to accumulate failures on nested objects.

For example, say we have the below json example where fields `nestedB` and `nestedC` is missing:

```scala
val j =
  """
    |{
    |  "a": 1,
    |  "b": 1,
    |  "c": {
    |    "nestedA": 2,
    |    "nestedD": 5
    |  }
    |}
  """.stripMargin

case class Inner(nestedA: Int, nestedB: Int, nestedC: Int, nestedD: Int)
case class Outter(a: Int, b: Int, c: Inner)

implicit val innerDecoder: Decoder[Inner] = (
  Decoder.instance(_.downField("nestedA").as[Int]),
  Decoder.instance(_.downField("nestedB").as[Int]),  // missing in above json
  Decoder.instance(_.downField("nestedC").as[Int]),  // missing in above json
  Decoder.instance(_.downField("nestedD").as[Int])
).mapN(Inner(_, _, _, _))

implicit val outterDecoder: Decoder[Outter] = (
  Decoder.instance(_.downField("a").as[Int]),
  Decoder.instance(_.downField("b").as[Int]),
  Decoder.instance(_.downField("c").as[Inner])
).mapN(Outter(_, _, _))

val Invalid(result) = io.circe.parser.decodeAccumulating[Outter](j)
```

I would expect that the `Invalid` result would contain a `Nel` with two entries, for each of the missing properties, but instead it short-circuits to only one:

```scala
NonEmptyList(DecodingFailure(Attempt to decode value on failed cursor, List(DownField(nestedB), DownField(c))))
```

**Comment by travisbrown** [comment]
@dr4ke616 Unfortunately the error accumulation mechanism is pretty fragile, in the sense that it only works when you compose and transform `Decoder` instances in certain ways. For example, if you use an accumulating decoder inside of `Decoder.instance`, you lose accumulation there.

This is specifically the case with `Decoder.instance` because it creates a decoder from an arbitrary `HCursor => Result[A]` method, so it's not possible (barring some horrible macro magic I don't want in circe-core) to know what the corresponding `HCursor => ValidationNel[Error, A]` should look like.

The following should do what you want, since it uses `prepare`:

```scala
case class Inner(nestedA: Int, nestedB: Int, nestedC: Int, nestedD: Int)
case class Outter(a: Int, b: Int, c: Inner)

implicit val innerDecoder: Decoder[Inner] = (
  Decoder.instance(_.downField("nestedA").as[Int]),
  Decoder.instance(_.downField("nestedB").as[Int]),  // missing in above json
  Decoder.instance(_.downField("nestedC").as[Int]),  // missing in above json
  Decoder.instance(_.downField("nestedD").as[Int])
).mapN(Inner(_, _, _, _))

implicit val outterDecoder: Decoder[Outter] = (
  Decoder.instance(_.downField("a").as[Int]),
  Decoder.instance(_.downField("b").as[Int]),
  Decoder[Inner].prepare(_.downField("c"))
).mapN(Outter(_, _, _))
```
(As a side note, `_.get[Int]("a")` is equivalent to `_.downField("a").as[Int]`.)

**Comment by dr4ke616** [comment]
Thanks, that makes sense.

Im still however seeing unexpected behaviour where if the `Inner` object is optional (`Option[Inner]` as apposed to just `Inner`, both in the defined case class and the decoder), it doesn't accumulate the errors like it does if it weren't optional.

For example:
```scala
val j =
  """
    |{
    |  "a": 1,
    |  "b": 1,
    |  "c": 1,
    |  "d": {
    |    "nestedA": 2,
    |    "nestedD": 5
    |  }
    |}
  """.stripMargin

case class Inner(nestedA: Int, nestedB: Int, nestedC: Int, nestedD: Int)
case class Outter(a: Int, b: Int, c: Int, d: Option[Inner])

implicit val innerDecoder: Decoder[Inner] = (
  Decoder.instance(_.get[Int]("nestedA")),
  Decoder.instance(_.get[Int]("nestedB")),
  Decoder.instance(_.get[Int]("nestedC")),
  Decoder.instance(_.get[Int]("nestedD"))
).mapN(Inner(_, _, _, _))

implicit val ss: Decoder[Outter] = (
  Decoder.instance(_.get[Int]("a")),
  Decoder.instance(_.get[Int]("b")),
  Decoder.instance(_.get[Int]("c")),
  Decoder[Option[Inner]].prepare(_.downField("d"))  // works as expected if it were just Inner
).mapN(Outter(_, _, _, _))

val result = io.circe.parser.decodeAccumulating[Outter](j)(ss)
```

This will result in:

```scala
NonEmptyList(DecodingFailure(Attempt to decode value on failed cursor, List(DownField(nestedB), DownField(d))))
```

Where if the `Inner` class weren't optional it would yield:

```scala
Invalid(NonEmptyList(DecodingFailure(Attempt to decode value on failed cursor, List(DownField(nestedB), DownField(d))), DecodingFailure(Attempt to decode value on failed cursor, List(DownField(nestedC), DownField(d)))))
```

**Comment by travisbrown** [comment]
Hmm, I'm not able to take a closer look right now, but am reopening the issue.

**Comment by travisbrown** [comment]
Fixed in #1153.
